### PR TITLE
CORDA-3871: New integration test for handshake timeout in AMQPClient

### PR DESCRIPTION
**Pull request by vkolomeyko**
This test verifies some edge case scenarios like handshake timeouts when `AMQPClient` connected to the server.
In order to have control over handshake internals a simple TLS server is created which may have a configurable handshake delay.

In order to prove that newly introduced tests are not flaky they were executed on CI 50 times, evidence [here](https://ci01.dev.r3.com/job/Corda-Open-Source/job/Corda-OS-Pull-Request-Mandatory-Tests/job/corda/job/PR-6519/6/testReport/net.corda.node.amqp/AMQPClientSslErrorsTest/amqpClientServerConnect_iteration___50_/).

### CODE REVIEW

File: node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+        serverThread.use {
+            val amqpClient = AMQPClient(listOf(NetworkHostAndPort("localhost", serverPort)), setOf(ALICE_NAME), clientAmqpConfig)
+
+            amqpClient.use {
+                val clientConnected = amqpClient.onConnection.toFuture()
+                amqpClient.start()
+                val clientConnect = clientConnected.get()
+                assertFalse(clientConnect.connected)
+                // Not a badCert, but a timeout during handshake
+                assertFalse(clientConnect.badCert)
```
Comment (vkolomeyko) [inline node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt:203]: Main condition to verify. Should not be bad certificate.
Verified that this test will fail if fix done earlier #6409 is reverted.

Comment (mnesbit) [reply node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt:203]: Looks pretty much as I hoped. Would quite value being able to stop at different points through the handshake, but not sure how practical that is.

Comment (vkolomeyko) [reply node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt:203]: We still rely on `javax.net.ssl.SSLEngine` and `sun.security.ssl.ServerHandshaker` to perform handshake, so cannot really introduce delays at `clientHello`, `serverHello`, `serverKeyExchange` etc phases. Since we are talking about introducing delay on the server side during handshake, I can think of the following phases we still have control over:
1.  BeginHandshake;
2. Optional delay that has been introduced;
3. NEED_WRAP
4. Could introduce anther delay here. But I do not see it making much of a difference as far as handshake timing is concerned.

File: node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+                assertTrue(clientConnect.connected)
+
+                log.info("Confirmed connected")
+            }
+        }
+        assertFalse(serverThread.isActive)
+    }
+
+    @Test(timeout = 300_000)
+    fun amqpClientServerHandshakeTimeout() {
```
Comment (rekalov) [inline node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt:190]: This test looks good. Can we add more tests for "close_notify" and "writing TLS control frames" causes?

File: node/src/integration-test/java/net/corda/node/amqp/ServerThread.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+/**
+ * This class provides a runnable that can be used to initialize a {@link NioSslServer} thread.
+ * <p/>
+ * Run starts the server, which will start listening to the configured IP address and port for
+ * new SSL/TLS connections and serve the ones already connected to it.
+ * <p/>
+ * Also a stop method is provided in order to gracefully close the server and stop the thread.
+ *
+ * @author <a href="mailto:alex.a.karnezis@gmail.com">Alex Karnezis</a>
```
Comment (rekalov) [inline node/src/integration-test/java/net/corda/node/amqp/ServerThread.java:19]: We incorporate (and probably modify) thirdparty code into Corda OS - did you check licenses etc?
Ideally, I would stay away from this approach and use libraries if possible.

Also, so we really need this code? My impression that we can achieve the same (at least for timeout check) by using AMQPServer, netty server or something else by overriding X509ExtendedTrustManager.

File: node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt (hunk 2 above)
Comment (vkolomeyko) [reply node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt:190]: I tried, but in current server side code it is not that easy to activate a condition which will force Netty Client code to throw exception containing `close_notify` or `writing TLS control frames`.
Also, if it will be possible to achieve such an effect - that test might become incredibly flaky.

SSL Handshake timeout was a low hanging fruit in this regard.

File: node/src/integration-test/java/net/corda/node/amqp/ServerThread.java (hunk 3 above)
Comment (vkolomeyko) [reply node/src/integration-test/java/net/corda/node/amqp/ServerThread.java:19]: Re: License
Yes, I have checked: https://github.com/alkarn/sslengine.example/blob/master/LICENSE.md
and we are allowed to modify this code. Especially given that we include it into OS project.

As for whether we need this code:
This is largely the matter of the approach selected. It is possible to see that in this PR I managed to avoid materially modifying **any** production classes, but still have required handshake timeout functionality tested by standing up a custom SSL server.

I agree that the other approach may be to use `AMQPServer` and in fact - this is easier. But that would require changing `AMQPServer` to allow injecting test code for introducing delays in "special" places.
As for overriding `X509ExtendedTrustManager` - It is enough to have a look at the places where `LoggingTrustManagerWrapper` is created.
If we go with this approach we will have to have a mechanism to add flexibility in that regard, so that we could inject custom implementation of `X509ExtendedTrustManager` in all those places.

@mnesbit - WDYT about current approach taken?

Comment (mnesbit) [reply node/src/integration-test/java/net/corda/node/amqp/ServerThread.java:19]: Need to credit the source (we have a page somewhere), but approach is the one I wanted for this

Comment (mnesbit) [reply node/src/integration-test/java/net/corda/node/amqp/ServerThread.java:19]: using AMQPServer is not going to work as it is same netty code as client you are trying to test. I prefer this independent check

Comment (rekalov) [reply node/src/integration-test/java/net/corda/node/amqp/ServerThread.java:19]: I'm fine as long as the source is credited.

File: node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt (hunk 2 above)
Comment (rekalov) [reply node/src/integration-test/kotlin/net/corda/node/amqp/AMQPClientSslErrorsTest.kt:190]: ok

**Review by mnesbit — APPROVED** [review]
Thanks
